### PR TITLE
SPLAT-1452: Changed vcenter MaxItems to be 3 for TechPreview.

### PR DESCRIPTION
**Pull request by vr4manta**
[SPLAT-1452](https://issues.redhat.com/browse/SPLAT-1452)

### Changes
- Created new VSphereMultiVCenters feature gate
- Generated new CRDs with max vcenter = 3 when FeatureGate "VCenterMultiVCenters" is active
- Default FG remains 1 max vcenter
- Added example of new FeatureGateAwareMaxItems

### Dependencies
- https://github.com/openshift/kubernetes-sigs-controller-tools/pull/19

### Blocks
- https://github.com/openshift/installer/pull/8221
- https://github.com/openshift/machine-config-operator/pull/4298
- https://github.com/openshift/cluster-storage-operator/pull/465
- https://github.com/openshift/vmware-vsphere-csi-driver-operator/pull/228

### CODE REVIEW

File: config/v1/types_infrastructure.go [hunk 1]
```diff
@@ -1343,8 +1343,9 @@ type VSpherePlatformSpec struct {
 	// ---
 	// + If VCenters is not defined use the existing cloud-config configmap defined
 	// + in openshift-config.
-	// +kubebuilder:validation:MaxItems=1
 	// +kubebuilder:validation:MinItems=0
+	// +openshift:validation:FeatureGateAwareMaxItems:featureGate="",maxItems=1
+	// +openshift:validation:FeatureGateAwareMaxItems:featureGate=VSphereMultiVCenters,maxItems=3
```
Comment (jcpowermac) [inline config/v1/types_infrastructure.go:1348]: @vr4manta just curious why only three? Is there a documented maximum?

Comment (vr4manta) [reply config/v1/types_infrastructure.go:1348]: Yes, is in the epic.  Not sure the reasoning for 3 being the max, but for now, that's what we have it being set to.

File: example/v1/types_stable.go [hunk 2]
```diff
@@ -67,6 +67,14 @@ type StableConfigTypeSpec struct {
 	// celUnion demonstrates how to validate a discrminated union using CEL
 	// +optional
 	CELUnion CELUnion `json:"celUnion,omitempty"`
+
+	// EvolvingCollection demonstrates how to have a collection where the maximum number of items varies on cluster type.
```
Comment (JoelSpeed) [inline example/v1/types_stable.go:71]: Nit, comments should start with json tag version
```suggestion
	// evolvingCollection demonstrates how to have a collection where the maximum number of items varies on cluster type.
```

File: config/v1/types_infrastructure.go (hunk 1 above)
Comment (JoelSpeed) [inline config/v1/types_infrastructure.go:1348]: Does this feature gate already exist?

Comment (vr4manta) [reply config/v1/types_infrastructure.go:1348]: We had the variable before where we capped it to 1 max.  We never supported multiple vCenters.  This feature gate is to allow us to provide that function and use the gate in multiple operators where today we are limiting vCenter count to 1.  In fact, multiple operators have hard coded limit set to one.  With this feature gate, we'll be updating those logic to now allow multiple if this new FeatureGate is enabled.

File: example/v1/types_stable.go (hunk 2 above)
Comment (vr4manta) [reply example/v1/types_stable.go:71]: Will fix.  Thanks!

File: config/v1/types_infrastructure.go (hunk 1 above)
Comment (JoelSpeed) [reply config/v1/types_infrastructure.go:1348]: The feature gate will need to be initialised in `features/features.go` else the tooling (and operators) won't actually know the feature gate exists

Comment (vr4manta) [reply config/v1/types_infrastructure.go:1348]: @JoelSpeed , yes, I updated that file in this PR already.  Is that the change you are thinking or did I miss something?

Comment (JoelSpeed) [reply config/v1/types_infrastructure.go:1348]: Apparently I just couldn't find it, ignore me

File: machineconfiguration/v1/tests/controllerconfigs.machineconfiguration.openshift.io/VSphereMultiVCenters.yaml [hunk 3]
```diff
@@ -0,0 +1,103 @@
+apiVersion: apiextensions.k8s.io/v1 # Hack because controller-gen complains if we don't have this
+name: "Shamefully missing"
```
Comment (JoelSpeed) [inline machineconfiguration/v1/tests/controllerconfigs.machineconfiguration.openshift.io/VSphereMultiVCenters.yaml:2]: Can we make this more appropriate to the feature/types, see how we format this for other tests

File: machineconfiguration/v1/tests/controllerconfigs.machineconfiguration.openshift.io/VSphereMultiVCenters.yaml [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+          network: null
+          networkType: OVNKubernetes
+          osImageURL: example.com/example/openshift-release-dev@sha256:eacdc37aec78fdbf8caa9601e4012ab31453cf59b086474901900e853e803ea8
+          platform: none
+          proxy: null
+          pullSecret:
+            name: pull-secret
+            namespace: openshift-config
+          releaseImage: ""
+          rootCAData: Y2VydGlmaWNhdGUK
```
Comment (JoelSpeed) [inline machineconfiguration/v1/tests/controllerconfigs.machineconfiguration.openshift.io/VSphereMultiVCenters.yaml:103]: Nit, missing a new line at the end here

File: machineconfiguration/v1/tests/controllerconfigs.machineconfiguration.openshift.io/VSphereMultiVCenters.yaml (hunk 3 above)
Comment (vr4manta) [reply machineconfiguration/v1/tests/controllerconfigs.machineconfiguration.openshift.io/VSphereMultiVCenters.yaml:2]: This test file confused me.  For some reason the tooling said I had to add it when I didn't make a direct change to it.  I copied the same test as VSphereControlPlaneMachineSet.yaml. In fact, reviewing all the other tests for controllerconfigs, they are the same copy.

Comment (vr4manta) [reply machineconfiguration/v1/tests/controllerconfigs.machineconfiguration.openshift.io/VSphereMultiVCenters.yaml]: I updated the test name the new convention.  We'll need to update the rest of the tests in that directory at a later time.

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

**Comment by JoelSpeed** [comment]
/override ci/prow/verify-crd-schema 

Pre-existing failures being added to the new FeatureGated version of the file

**Comment by JoelSpeed** [comment]
/override ci/prow/verify-crd-schema 

Pre-existing failures, see previous

**Pull request merged**